File: authorization-for-rag/langchain4j-java/src/main/java/rag/FGAInit.java
package rag;

import com.fasterxml.jackson.databind.ObjectMapper;
import dev.openfga.sdk.api.client.OpenFgaClient;
import dev.openfga.sdk.api.client.model.ClientTupleKey;
import dev.openfga.sdk.api.client.model.ClientWriteRequest;
import dev.openfga.sdk.api.configuration.ClientConfiguration;
import dev.openfga.sdk.api.configuration.ClientCredentials;
import dev.openfga.sdk.api.configuration.ClientWriteOptions;
import dev.openfga.sdk.api.configuration.Credentials;
import dev.openfga.sdk.api.model.CreateStoreRequest;
import dev.openfga.sdk.api.model.WriteAuthorizationModelRequest;
import io.github.cdimascio.dotenv.Dotenv;

import java.util.List;

public class FGAInit {
    static {
        Dotenv.configure().systemProperties().load();
    }

    public static void main(String[] args) throws Exception {
        // create a new FGA store
        var config = new ClientConfiguration()
                .apiUrl(System.getProperty("FGA_API_URL", "http://localhost:8080"))
                // Credentials required only for Auth0 FGA
                .credentials(new Credentials(new ClientCredentials()
                        .apiTokenIssuer(System.getProperty("FGA_API_TOKEN_ISSUER", "auth.fga.dev"))
                        .apiAudience(System.getProperty("FGA_API_AUDIENCE", "https://api.us1.fga.dev/"))
                        .clientId(System.getProperty("FGA_CLIENT_ID"))
                        .clientSecret(System.getProperty("FGA_CLIENT_SECRET"))));

        var fgaClient = new OpenFgaClient(config);
        var body = new CreateStoreRequest().name("FGA Demo Store");
        var store = fgaClient.createStore(body).get();
        fgaClient.setStoreId(store.getId());

        System.out.println("FGA Store ID: " + store.getId());

        // Create FGA model
        var schema = """
                {
                  "schema_version":"1.1",
                  "type_definitions": [
                    {
                      "type":"user"
                    },
                    {
                      "metadata": {
                        "relations": {
                          "owner": {
                            "directly_related_user_types": [
                              {
                                "type":"user"
                              }
                            ]
                          },
                          "viewer": {
                            "directly_related_user_types": [
                              {
                                "type":"user"
                              },
                              {
                                "type":"user",
                                "wildcard": {}
                              }
                            ]
                          }
                        }
                      },
                      "relations": {
                        "owner": {
                          "this": {}
                        },
                        "viewer": {
                          "this": {}
                        }
                      },
                      "type":"doc"
                    }
                  ]
                }
                """;

        var mapper = new ObjectMapper().findAndRegisterModules();
        var authorizationModel = fgaClient
                .writeAuthorizationModel(mapper.readValue(schema, WriteAuthorizationModelRequest.class))
                .get();

        System.out.println("Model created: " + authorizationModel.getAuthorizationModelId());


        // Create Tuples
        var request = new ClientWriteRequest()
                .writes(List.of(
                        new ClientTupleKey()
                                .user("user:*")
                                .relation("viewer")
                                ._object("doc:public-doc"),
                        new ClientTupleKey()
                                .user("user:user1")
                                .relation("viewer")
                                ._object("doc:private-doc")
                ));

        var options = new ClientWriteOptions()
                .authorizationModelId(authorizationModel.getAuthorizationModelId())
                .disableTransactions(false);

        var response = fgaClient.write(request, options).get();

        System.out.println("Tuples created: Status " + response.getStatusCode());
    }
}

File: authorization-for-rag/openai-fga-python/scripts/fga_init.py
"""
This initializes the FGA store with the necessary tuple data
it will use the openfga_sdk and read the configuration from the .config file
"""
import asyncio
# import the configuration from the helpers/config.py file
from helpers.config import config
from openfga_sdk import ClientConfiguration, OpenFgaClient
from openfga_sdk.credentials import Credentials, CredentialConfiguration
from openfga_sdk.client.models import ClientTuple, ClientWriteRequest


async def fga_setup(config):
    fga_config = ClientConfiguration(
        api_url=config["AUTH0FGA"]["FGA_API_URL"],
        store_id=config["AUTH0FGA"]["FGA_STORE_ID"],
        credentials=Credentials(
            method="client_credentials",
            configuration=CredentialConfiguration(
                api_issuer=config["AUTH0FGA"]["FGA_API_TOKEN_ISSUER"],
                api_audience=config["AUTH0FGA"]["FGA_API_AUDIENCE"],
                client_id=config["AUTH0FGA"]["FGA_CLIENT_ID"],
                client_secret=config["AUTH0FGA"]["FGA_CLIENT_SECRET"],
            )
        )
    ) 

    fga_client = OpenFgaClient(fga_config)
    return fga_client


async def main():
    fga_client = await fga_setup(config)
    tuple_data = {"user":"user:jess", "relation":"viewer", "object":"doc:public-doc"}
    body = ClientWriteRequest(writes=[ ClientTuple(**tuple_data) ])
    await fga_client.write(body)
    await fga_client.close()


if __name__ == "__main__":
    asyncio.run(main())